packages/permission-log-controller/src/PermissionLog.test.ts

```diff
@@ -0,0 +1,792 @@
+import { ControllerMessenger } from '@metamask/base-controller';
+import type {
+  JsonRpcEngineReturnHandler,
+  JsonRpcEngineNextCallback,
+} from '@metamask/json-rpc-engine';
+import type { PendingJsonRpcResponse } from '@metamask/utils';
+import { nanoid } from 'nanoid';
+import { useFakeTimers } from 'sinon';
+import type { SinonFakeTimers } from 'sinon';
+
+import { LOG_LIMIT, LOG_METHOD_TYPES } from './enums';
+import { PermissionLogController } from './PermissionLog';
+import type {
+  JsonRpcRequestWithOrigin,
+  Permission,
+  PermissionActivityLog,
+} from './PermissionLog';
+import { CaveatTypes, constants, getters, noop } from './permissions';
+
+const { PERMS, RPC_REQUESTS } = getters;
+const {
+  ACCOUNTS,
+  EXPECTED_HISTORIES,
+  SUBJECTS,
+  PERM_NAMES,
+  REQUEST_IDS,
+  RESTRICTED_METHODS,
+} = constants;
+
+let clock: SinonFakeTimers;
+
+type LogMiddleware = (
+  req: JsonRpcRequestWithOrigin,
+  res: PendingJsonRpcResponse<(string | Permission)[]>,
+  next?: JsonRpcEngineNextCallback,
+) => void;
+
+const name = 'PermissionLogController';
+
+/**
+ * Constructs a restricted controller messenger.
+ *
+ * @returns A restricted controller messenger.
+ */
+function getMessenger() {
+  return new ControllerMessenger().getRestricted<typeof name, never, never>({
+    name,
+  });
+}
+
+const initPermLog = (state = {}): PermissionLogController => {
+  const messenger = getMessenger();
+  return new PermissionLogController({
+    messenger,
+    restrictedMethods: RESTRICTED_METHODS,
+    state,
+  });
+};
+
+const mockNext: JsonRpcEngineNextCallback = (handler) => {
+  if (handler) {
+    handler(noop);
+  }
+};
+
+const initMiddleware = (permLog: PermissionLogController): LogMiddleware => {
+  const middleware = permLog.createMiddleware();
+  return (
+    req: JsonRpcRequestWithOrigin,
+    res: PendingJsonRpcResponse<(string | Permission)[]>,
+    next = mockNext,
+  ) => {
+    middleware(req, res, next);
+  };
+};
```

Comment:
Following the comment above, this is how it would look if we used `JsonRpcMiddleware`:

``` suggestion
const initMiddleware = (
  permLog: PermissionLogController,
): JsonRpcMiddleware<JsonRpcParams, Json> => {
  const middleware = permLog.createMiddleware();
  return (req, res, next = mockNext, end = noop) => {
    middleware(req, res, next, end);
  };
};